### PR TITLE
MBS-9536 Fix negative state assertions

### PR DESCRIPTION
**Pull request by RuslanMingaliev**

### CODE REVIEW

File: subprojects/common/waiter/src/main/kotlin/com/avito/android/util/Waiter.kt [hunk 1]
```diff
@@ -13,5 +14,15 @@ fun <T> waitForAssertion(
     action = action
 )
 
-private const val DEFAULT_TIMEOUT_SEC = 5000L
+fun continuousAssertion(
```
Comment (sergioboi) [inline subprojects/common/waiter/src/main/kotlin/com/avito/android/util/Waiter.kt:17]: : )))

File: subprojects/common/waiter/src/main/kotlin/com/avito/android/util/Waiter.kt [hunk 2]
```diff
@@ -1,9 +1,10 @@
 package com.avito.android.util
 
+import com.avito.android.waiter.repeatFor
 import com.avito.android.waiter.waitFor
 
 fun <T> waitForAssertion(
-    timeoutMilliseconds: Long = DEFAULT_TIMEOUT_SEC,
+    timeoutMilliseconds: Long = DEFAULT_TIMEOUT_MS,
```
Comment (sergioboi) [inline subprojects/common/waiter/src/main/kotlin/com/avito/android/util/Waiter.kt:7]: Are 5 seconds not too much?

Comment (RuslanMingaliev) [reply subprojects/common/waiter/src/main/kotlin/com/avito/android/util/Waiter.kt:7]: Thought about it. Don't know how to choose an optimal timeout

File: subprojects/common/waiter/src/main/kotlin/com/avito/android/waiter/Waiter.kt [hunk 3]
```diff
@@ -37,3 +37,19 @@ fun <T> waitFor(
...
+    sleepAction: (frequencyMs: Long) -> Unit = { Thread.sleep(it) },
+    action: () -> Unit
+) {
+    var timer = 0L
+    val startTime = System.currentTimeMillis()
+
+    do {
+        action.invoke()
+        sleepAction(frequencyMs)
+        timer += frequencyMs
```
Comment (eugene-krivobokov) [inline subprojects/common/waiter/src/main/kotlin/com/avito/android/waiter/Waiter.kt]: If I'm not mistaken, "timer" can be deleted.
The check `System.currentTimeMillis() - startTime <= timeoutMs` already do the job.

File: subprojects/common/waiter/src/main/kotlin/com/avito/android/util/Waiter.kt (hunk 2 above)
Comment (dsvoronin) [reply subprojects/common/waiter/src/main/kotlin/com/avito/android/util/Waiter.kt:7]: only by collecting enough data.
Do we have some metrics about it?

File: subprojects/common/waiter/src/main/kotlin/com/avito/android/waiter/Waiter.kt (hunk 3 above)
Comment (RuslanMingaliev) [reply subprojects/common/waiter/src/main/kotlin/com/avito/android/waiter/Waiter.kt]: Yes, deleted it from the `waitFor` function too.

File: subprojects/common/waiter/src/main/kotlin/com/avito/android/util/Waiter.kt (hunk 2 above)
Comment (RuslanMingaliev) [reply subprojects/common/waiter/src/main/kotlin/com/avito/android/util/Waiter.kt:7]: No, we don't. I'll add it if you help me with the correct series name and Graphana visualization.

File: subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/ActivityLifecycleStatesTest.kt [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    @get:Rule
+    val activityRule = inHouseScreenRule<EmptyActivity>()
+
+    @Before
+    fun before() {
+        activityRule.launchActivity(null)
+    }
+
+    @Test
+    fun finishedActivity_isInDestroyedState_succeed() {
```
Comment (eugene-krivobokov) [inline subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/ActivityLifecycleStatesTest.kt]: We describe another test name pattern in conventions: https://avito-tech.github.io/avito-android/contributing/Testing/#naming-conventions
`state under test (SUT) - expected behaviour - [when]`

I evaluated parts as:

- finishedActivity - when
- isInDestroyedState - SUT
- succeed - expected behaviour

According to this pattern, it should be `isInDestroyedState - succeed - finishedActivity`

File: subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/ActivityLifecycleStatesTest.kt [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    @Test
+    fun finishedActivity_isNotInDestroyedState_throwsAssertionError() {
+        activityRule.scenario.onActivity { it.finish() }
+        assertThrows<AssertionError> { activityRule.checks.isNotInState(Lifecycle.State.DESTROYED) }
+    }
+
+    @Test
+    fun launchedActivity_isNotInDestroyedState_succeed() {
```
Comment (eugene-krivobokov) [inline subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/ActivityLifecycleStatesTest.kt]: `isNotInDestroyedState` --> `isNotDestroyed` ?
It looks enough to express the idea.

File: subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/ActivityLifecycleStatesTest.kt [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+import ru.avito.util.assertThrows
+
+class ActivityLifecycleStatesTest {
+
+    @get:Rule
+    val activityRule = inHouseScreenRule<EmptyActivity>()
+
+    @Before
+    fun before() {
+        activityRule.launchActivity(null)
```
Comment (eugene-krivobokov) [inline subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/ActivityLifecycleStatesTest.kt]: What about cases when activity was not launched?

File: subprojects/common/waiter/src/main/kotlin/com/avito/android/util/Waiter.kt (hunk 2 above)
Comment (dsvoronin) [reply subprojects/common/waiter/src/main/kotlin/com/avito/android/util/Waiter.kt:7]: that will be a lot of data. What do we need it for? If only to determine this value i see it as an overkill

File: subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/ActivityLifecycleStatesTest.kt (hunk 6 above)
Comment (RuslanMingaliev) [reply subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/ActivityLifecycleStatesTest.kt]: The test will [fail](https://github.com/avito-tech/avito-android/blob/dbdb289b63394d7dcce0a2044445c443dc2830dd/subprojects/test-runner/test-inhouse-runner/src/main/kotlin/com/avito/android/rule/base/BaseActivityScenarioRule.kt#L18) then

File: subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/ActivityLifecycleStatesTest.kt (hunk 4 above)
Comment (RuslanMingaliev) [reply subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/ActivityLifecycleStatesTest.kt]: Yes, I didn't pay attention

File: subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/ActivityLifecycleStatesTest.kt (hunk 6 above)
Comment (eugene-krivobokov) [reply subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/ActivityLifecycleStatesTest.kt]: Ok, should we test it here?

**Review by eugene-krivobokov — APPROVED** [review]

File: subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/ActivityLifecycleStatesTest.kt (hunk 6 above)
Comment (RuslanMingaliev) [reply subprojects/android-test/ui-testing-core-app/src/androidTest/kotlin/com/avito/android/ui/test/ActivityLifecycleStatesTest.kt]: It never hurts to check, so added a new test.

**Pull request merged**